adding all missing types of two qubit SWAP gates and sSQRTZZ/sInvSQRTZZ gate

Fe-r-oz opened the pull request:
This PR aims to add new types of `SWAP` gates.

Reference: https://github.com/quantumlib/Stim/blob/main/doc/gates.md

Exact notation as stim uses has been used. There were many 2 qubit gates, so I decided to go small steps at a time. 

Thank you for reviweing the PR. 

```
julia> sCXSWAP(1,2)
sCXSWAP on qubit1 (1,2)
X₁ ⟼ + XX
X₂ ⟼ + X_
Z₁ ⟼ + _Z
Z₂ ⟼ + ZZ

julia> sCZSWAP(1,2)
sCZSWAP on qubit1 (1,2)
X₁ ⟼ + ZX
X₂ ⟼ + XZ
Z₁ ⟼ + _Z
Z₂ ⟼ + Z_

julia> sSWAPCX(1,2)
sSWAPCX on qubit1 (1,2)
X₁ ⟼ + _X
X₂ ⟼ + XX
Z₁ ⟼ + ZZ
Z₂ ⟼ + Z_

julia> sSWAPCXDAG(1,2)
sSWAPCXDAG on qubit1 (1,2)
X₁ ⟼ + XX
X₂ ⟼ + X_
Z₁ ⟼ + _Z
Z₂ ⟼ + ZZ

```

f you want to submit an unfinished piece of work in order to get comments and discuss, please mark the pull request as a draft and ping the repository maintainer.

**Please address only one topic or issue per pull request! Many small PRs are much easier to review and merge than one large PR.**

Before merging, all changes and new functionality should be marked in the CHANGELOG file, but feel free to just leave your CHANGELOG notes in the PR description, to avoid merge conflicts with other requests modifying that file. The maintainer will add these CHANGELOG notes for you if you do so.

Before considering your pull request ready for review and merging make sure that all of the following are completed (please keep the clecklist as part of your PR):

- [x] The code is properly formatted and commented.
- [x] All new functionality is tested.
- [x] All of the automated tests on github pass.


If possible, keep your git history not too wild (rebase and squash commits, keep commits small and semantically separated) so that review is easier.

Krastanov reviewed (changes requested):
This looks great! Same comments as in #333

Fe-r-oz commented:
Added `ISWAP` and it's inverse as well so we now have all types of  two qubit `SWAP` gates w.r.t to stim. 
Also, added `SQRTZZ` and it's inverse for now! Followed your task todo list for these latest additions as well!

Krastanov commented on test/test_symcliff.jl:
```diff
@@ -88,6 +88,19 @@
             @test CliffordOperator(inv(sXCZ(n₁, n₂)), n₁) == inv(CliffordOperator(sCNOT(n₂, n₁), n₁))
             @test CliffordOperator(inv(sZCrY(n₁, n₂)), n₁) == inv(CliffordOperator(sZCrY(n₁, n₂), n₁))
             @test CliffordOperator(inv(sInvZCrY(n₁, n₂)), n₁) == inv(CliffordOperator(sInvZCrY(n₁, n₂), n₁))
+            @test CliffordOperator(inv(sCXSWAP(n₁, n₂)), n₁) == inv(CliffordOperator(sInvSWAPCX(n₁, n₂), n₁))
```
Where does this come from?

Fe-r-oz replied:
`CXSWAP`: https://github.com/quantumlib/Stim/blob/main/doc/gates.md#CXSWAP

`SWAPCX:` https://github.com/quantumlib/Stim/blob/main/doc/gates.md#SWAPCX

I used the` inv(CliffordOperator(...))` to find `SWAPCX`  inverse. I saw that `InvSWAPCX` is the same as` CXSWAP`.  So, I added a test for this relationship as well.

Krastanov commented on test/test_symcliff.jl:
```diff
@@ -88,6 +88,19 @@
+    @testset "New two qubit gates consistency check" begin
+        # see https://github.com/quantumlib/Stim/blob/main/doc/gates.md
+        @test CliffordOperator(sSWAPCX)    == C"IX XX ZZ ZI"
+        @test CliffordOperator(sInvSWAPCX) == C"XX XI IZ ZZ"
```
This does not seem to be in the linked list of gates. Where is it from?

Fe-r-oz replied:
`SWAPCX:` https://github.com/quantumlib/Stim/blob/main/doc/gates.md#SWAPCX

I used the` inv(CliffordOperator(...))` to find it's inverse.

Krastanov commented on test/test_symcliff.jl:
```diff
@@ -88,6 +88,19 @@
+        @test CliffordOperator(sCXSWAP)    == C"XX XI IZ ZZ"
+        @test CliffordOperator(sISWAP)     == C"ZY YZ IZ ZI"
+        @test CliffordOperator(sInvISWAP)  == C"-ZY -YZ IZ ZI"
+        @test CliffordOperator(sSQRTZZ)    == C"YZ ZY ZI IZ"
+        @test CliffordOperator(sInvSQRTZZ) == C"-YZ -ZY ZI IZ"
```
Where are these two coming from?

Fe-r-oz replied:
`SQRTZZ: `https://github.com/quantumlib/Stim/blob/main/doc/gates.md#SQRT_ZZ
`SQRTZZDAG: `https://github.com/quantumlib/Stim/blob/main/doc/gates.md#the-sqrt_zz_dag-gate

Krastanov commented on src/symbolic_cliffords.jl:
```diff
@@ -347,20 +360,28 @@ function Base.show(io::IO, op::AbstractTwoQubitOperator)
+LinearAlgebra.inv(op::sCXSWAP)    = sSWAPCX(op.q1, op.q2)
+LinearAlgebra.inv(op::sISWAP)     = sInvISWAP(op.q1, op.q2)
+LinearAlgebra.inv(op::sInvISWAP)  = sISWAP(op.q1, op.q2)
+LinearAlgebra.inv(op::sSQRTZZ)    = sInvSQRTZZ(op.q1, op.q2)
+LinearAlgebra.inv(op::sInvSQRTZZ) = sSQRTZZ(op.q1, op.q2)
```
repeated question: where are these two from

Fe-r-oz replied:
Please see: https://github.com/QuantumSavory/QuantumClifford.jl/pull/336#discussion_r1759677752

Krastanov commented on test/test_symcliff.jl:
```diff
@@ -88,6 +88,19 @@
         end
     end
+
+    @testset "New two qubit gates consistency check" begin
```
```suggestion
    @testset "Consistency check with STIM conventions" begin
```

Krastanov reviewed (changes requested):
This would be a pretty nice addition, thank you! I have a few questions on where some of the definitions are coming from.

Fe-r-oz commented:
Hopefully, my comments were satisfactory. All the gate definitions are from stim [gates page](https://github.com/quantumlib/Stim/blob/main/doc/gates.md).

I think PR is ready for review, Thank you!

Krastanov commented:
Thank you! Great to have all of these!
